File: content-solutions/graph-advanced/mst-power-plants.js
class DSU {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.size = new Array(n).fill(1);
  }
  find(x) {
    let root = x;
    while (this.parent[root] !== root) root = this.parent[root];
    while (x !== root) {
      const next = this.parent[x];
      this.parent[x] = root;
      x = next;
    }
    return root;
  }
  union(a, b) {
    let ra = this.find(a);
    let rb = this.find(b);
    if (ra === rb) return false;
    if (this.size[ra] < this.size[rb]) [ra, rb] = [rb, ra];
    this.parent[rb] = ra;
    this.size[ra] += this.size[rb];
    return true;
  }
}

function solution(n, plants, cables) {
  const dsu = new DSU(n);
  for (const p of plants) dsu.union(plants[0], p);
  let total = 0;
  for (const [a, b, c] of [...cables].sort((x, y) => x[2] - y[2])) {
    if (dsu.union(a, b)) total += c;
  }
  return total;
}

File: content-solutions/graph-advanced/mst-power-plants.py
def solution(n, plants, cables):
    parent = list(range(n))
    size = [1] * n

    def find(x):
        root = x
        while parent[root] != root:
            root = parent[root]
        while x != root:
            nxt = parent[x]
            parent[x] = root
            x = nxt
        return root

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra == rb:
            return False
        if size[ra] < size[rb]:
            ra, rb = rb, ra
        parent[rb] = ra
        size[ra] += size[rb]
        return True

    for p in plants:
        union(plants[0], p)
    total = 0
    for a, b, c in sorted(cables, key=lambda x: x[2]):
        if union(a, b):
            total += c
    return total
